fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -419,6 +421,80 @@ void testAgilityContextAtomicAttribute(boolean useAgile) {
         }
     }
 
+    @Test
+    void testAgilityContextRecoveryPath() {
+        AtomicInteger refInt = new AtomicInteger(0);
+        final byte[] keyAborted;
+        final Tuple value = Tuple.from(800, "Green eggs and ham", 0);
+        byte[] packedValue = value.pack();
+        try (FDBRecordContext context = openContext()) {
+            final AgilityContext agilityContext = getAgilityContext(context, true);
+            final Subspace subspace = path.toSubspace(context);
+            keyAborted = subspace.pack(Tuple.from(2023, 3));
+            agilityContext.set(keyAborted, packedValue);
+            agilityContext.abortAndClose();
+            IntStream range = IntStream.rangeClosed(1, 10);
+            range.parallel().forEach(i -> {
+                if (i == 7) {
+                    agilityContext.applyInRecoveryPath(aContext -> {
+                        Assertions.assertNotNull(aContext);
+                        refInt.addAndGet(100);
+                        return CompletableFuture.completedFuture(null);
+                    }).join();
+                } else {
+                    Assertions.assertThrows(RecordCoreStorageException.class,
+                            () -> agilityContext.apply(aContext -> {
+                                refInt.set(i);
+                                return CompletableFuture.completedFuture(null);
+                            }).join());
+                }
+            });
+            assertEquals(100, refInt.get());
+            context.commit();
+        }
+    }
+
+    @Test
+    void testAgilityContextRecoveryPath2() {
+        final byte[] keyAborted;
+        final byte[] keySucceeds;
+        final byte[] keyFails;
+        final Tuple value = Tuple.from(800, "Green eggs and ham", 0);
+        byte[] packedValue = value.pack();
+        try (FDBRecordContext context = openContext()) {
+            final AgilityContext agilityContext = getAgilityContext(context, true);
+            final Subspace subspace = path.toSubspace(context);
+            keyAborted = subspace.pack(Tuple.from(2023, 3));
+            agilityContext.set(keyAborted, packedValue);
+            agilityContext.abortAndClose();
+            IntStream range = IntStream.rangeClosed(1, 10);
+            keySucceeds = this.path.toSubspace(context).pack(Tuple.from(prefix, "yes").pack());
+            keyFails = this.path.toSubspace(context).pack(Tuple.from(prefix, "no").pack());
```

Comment:
`path.toSubspace` can interact with the context. Specifically it may need to update the directory layer to do string interning.

It would be better to do this conversion in a transaction before to ensure that anything in the directory layer is committed beforehand.

Reply:
✅